cmd/kubeadm/app/util/config/common_test.go

```diff
@@ -213,209 +216,221 @@ func TestVerifyAPIServerBindAddress(t *testing.T) {
+				constants.InitConfigurationKind,
+				constants.ClusterConfigurationKind,
+				constants.JoinConfigurationKind,
+			},
+			expectErr: false,
+		},
+		{
+			desc: "Init + Cluster + Join Configurations are migrated",
+			oldCfg: dedent.Dedent(fmt.Sprintf(`
+			apiVersion: %s
+			kind: InitConfiguration
+			---
+			apiVersion: %[1]s
+			kind: ClusterConfiguration
+			---
+			apiVersion: %[1]s
+			kind: JoinConfiguration
+			discovery:
+			  bootstrapToken:
+			    token: abcdef.0123456789abcdef
+			    apiServerEndpoint: kube-apiserver:6443
+			    unsafeSkipCAVerification: true
+			`, gv)),
+			expectedKinds: []string{
+				constants.InitConfigurationKind,
+				constants.ClusterConfigurationKind,
+				constants.JoinConfigurationKind,
+			},
+			expectErr: false,
+		},
+		{
+			desc: "component configs are not migrated",
+			oldCfg: dedent.Dedent(fmt.Sprintf(`
+			apiVersion: %s
+			kind: InitConfiguration
+			---
+			apiVersion: %[1]s
+			kind: ClusterConfiguration
+			---
+			apiVersion: %[1]s
+			kind: JoinConfiguration
+			discovery:
+			  bootstrapToken:
+			    token: abcdef.0123456789abcdef
+			    apiServerEndpoint: kube-apiserver:6443
+			    unsafeSkipCAVerification: true
+			---
+			apiVersion: kubeproxy.config.k8s.io/v1alpha1
+			kind: KubeProxyConfiguration
+			---
+			apiVersion: kubelet.config.k8s.io/v1beta1
+			kind: KubeletConfiguration
+			`, gv)),
+			expectedKinds: []string{
+				constants.InitConfigurationKind,
+				constants.ClusterConfigurationKind,
+				constants.JoinConfigurationKind,
+			},
+			expectErr: false,
+		},
+		{
+			desc: "ClusterConfiguration gets migrated from experimental API",
+			oldCfg: dedent.Dedent(fmt.Sprintf(`
+			apiVersion: %s
+			kind: ClusterConfiguration
+			`, gv)),
+			expectedKinds: []string{
+				constants.InitConfigurationKind,
+				constants.ClusterConfigurationKind,
+			},
+		},
+	}
 
-// 	for _, test := range tests {
-// 		t.Run(test.desc, func(t *testing.T) {
-// 			b, err := MigrateOldConfig([]byte(test.oldCfg))
-// 			if test.expectErr {
-// 				if err == nil {
-// 					t.Fatalf("unexpected success:\n%s", b)
-// 				}
-// 			} else {
-// 				if err != nil {
-// 					t.Fatalf("unexpected failure: %v", err)
-// 				}
-// 				gvks, err := kubeadmutil.GroupVersionKindsFromBytes(b)
-// 				if err != nil {
-// 					t.Fatalf("unexpected error returned by GroupVersionKindsFromBytes: %v", err)
-// 				}
-// 				if len(gvks) != len(test.expectedKinds) {
-// 					t.Fatalf("length mismatch between resulting gvks and expected kinds:\n\tlen(gvks)=%d\n\tlen(expectedKinds)=%d",
-// 						len(gvks), len(test.expectedKinds))
-// 				}
-// 				for _, expectedKind := range test.expectedKinds {
-// 					if !kubeadmutil.GroupVersionKindsHasKind(gvks, expectedKind) {
-// 						t.Fatalf("migration failed to produce config kind: %s", expectedKind)
-// 					}
-// 				}
-// 			}
-// 		})
-// 	}
-// }
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			const allowExperimental = true
```

Comment:
Do we also need to test the scenarios where `allowExperimental` is false?
How about adding `allowExperimental bool` as an input variable of the test cases?

Reply:
added `allowExperimental bool`

this was actually wrong:
> Migration of experimental APIs is already tested in the config_test.go file in the "cmd" package and not in this test.

experimental API migration is now tested only in this test function in terms of output kinds.
TestValidateSupportedVersion also tests if the input API is allowed as experimental in general.

our tests are a bit sparce between migrate and validate, but i think we should try to keep it simple and not update many files / tests cases when e.g. v1beta4 is no longer experimental.

we could remove the tests from cmd/config_test.go and do all the migrate / validate tests in common_test.go on the backend layer.

Reply:
i can send a PR follow up.
it can remove the cmd/config_tests.go tests and add test for ValidateConfig in common_test.go. also see if more cases are needed for MigrateOldConfig and add them in common_test.go

Reply:
> i can send a PR follow up.

actually, seems easier better as a separate commit here.
will update this PR a bit later.

Reply:
added second commit